statsmodels/robust/covariance.py

```diff
@@ -0,0 +1,607 @@
+        Shrinkage for the scatter estimate. If it is zero, then no shrinkage
+        is performed. If it is None, then the shrinkage factor will be
+        determined by a plugin estimator
+    maxiter : int
+        maximum number of iterations to find the solution
+    eps : float
+        convergence criterion. The maximum absolute distance needs to be
+        smaller than eps for convergence.
+
+    Returns
+    -------
+    scatter : ndarray
+        estimate of the scatter matrix
+    iter : int
+        number of iterations used in finding a solution. If iter is less than
+        maxiter, then the iteration converged.
+    shrinkage_factor : float
+        shrinkage factor that was used in the estimation. This will be the
+        same as the function argument if it was not None.
+
+    Notes
+    -----
+    If the shrinkage factor is None, then a plugin is used as described in
+    Chen and Wiesel 2011. The required trace for a pilot scatter estimate is
+    obtained by the covariance rescaled by MAD estimate for the variance.
+
+    """
+    x = data_iterator
+    #x = np.asarray(data)
+    #nobs, k_vars = x.shape
+
+    # calculate MAD only once if needed
+    if start_cov is None or shrinkage_factor is None:
+        scale_mad = sm.robust.mad(x, center=0)
+
+    corr = None
+    if shrinkage_factor is None:
+        # maybe some things here are redundant
+        xd = x / x.std(0) #scale_mad
+        corr = xd.T.dot(xd)
+        corr * np.outer(scale_mad, scale_mad)
+        corr *= k_vars / np.trace(corr)
+        tr = np.trace(corr.dot(corr))
+
+        n, k = nobs, k_vars
+        # Chen and Wiesel 2011 equation (13)
+        sf = k*k + (1 - 2./k) * tr
+        sf /= (k*k - n*k - 2*n) + (n + 1 + 2. * (n - 1.) / k) * tr
+        shrinkage_factor = sf
+
+    if start_cov is not None:
+        c = start_cov
+    else:
+        c = np.diag(scale_mad**2)
+
+    identity = np.eye(k_vars)
+    kn = k_vars * 1. / nobs
+    for i in range(maxiter):
+        c_inv = np.linalg.pinv(c)
+        c_old = c
+        # this could be vectorized but could use a lot of memory
+        # TODO:  try to work in vectorized batches
+        # weights is a problem if iterator should be ndarray
+        #c0 = kn * sum(np.outer(xi, xj) / np.inner(xi, c_inv.dot(xj)) for xi, xj in x)
+        c0 = kn * sum(np.outer(xij[0], xij[1]) / np.inner(xij[0], c_inv.dot(xij[1])) for xij in x)
+        if shrinkage_factor != 0:
+            c = (1 - shrinkage_factor) * c0 + shrinkage_factor * identity
+        else:
+            c = c0
+
+        c *= k_vars / np.trace(c)
+
+        diff = np.max(np.abs(c - c_old))
+        if diff < eps:
+            break
+
+    return c, i, shrinkage_factor#, corr
+
+
+### iterative, M-estimators and related
+
+def cov_weighted(data, weights, center=None):
+    wsum = weights.sum()
+    if center is None:
+        wmean = weights.dot(data) / wsum
+    xdm = data - wmean
+    wcov = (weights * xdm.T).dot(xdm) / wsum
```

Comment:
make denominator optional   
(different function in general M-estimators or estimating equations, 
normalized for "shape" matrix by `det(wcov)*(1/k_vars)`